File: modules/shared/src/main/scala/shared/utils/StreamAppUtils.scala
package shared.utils

import java.util.Properties

import scala.collection.JavaConverters._
import org.apache.kafka.clients.admin.{AdminClient, CreateTopicsResult, NewTopic}
import org.apache.kafka.clients.consumer.ConsumerConfig
import org.apache.kafka.streams.errors.LogAndContinueExceptionHandler
import org.apache.kafka.streams.{KafkaStreams, StreamsConfig, Topology}
import shared.topics.TopicCreation
final case class StreamAppConfig(appId: String, bootstrapServers: String)

object StreamAppUtils {
  def getConfig(appConfig: StreamAppConfig): Properties = {
    val config: Properties = new Properties
    config.put(StreamsConfig.APPLICATION_ID_CONFIG, appConfig.appId)
    config.put(StreamsConfig.BOOTSTRAP_SERVERS_CONFIG, appConfig.bootstrapServers)
    config.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest")
    config.put(StreamsConfig.DEFAULT_DESERIALIZATION_EXCEPTION_HANDLER_CLASS_CONFIG,
               classOf[LogAndContinueExceptionHandler])
    config.put(StreamsConfig.PROCESSING_GUARANTEE_CONFIG, StreamsConfig.EXACTLY_ONCE)
    config
  }

  def addMissingTopics(adminClient: AdminClient)(topics: List[TopicCreation]): CreateTopicsResult = {
    val existingTopics = adminClient.listTopics().names().get().asScala
    val newTopics = topics.toSet
      .filter(t => !existingTopics.contains(t.topicName))
      .map(t => {
        val spec = t.topicSpec
        val newTopic =
          new NewTopic(t.topicName, spec.partitionCount(), spec.replicaCount()).configs(spec.config())
        newTopic
      })
    adminClient.createTopics(newTopics.asJava)
  }

  def runStreamApp(config: Properties, topology: Topology) = {
    val streams = new KafkaStreams(topology, new StreamsConfig(config))

    streams.cleanUp()
    streams.start()

    Runtime.getRuntime.addShutdownHook(new Thread {
      override def run(): Unit = {
        streams.close()
      }
    })
  }
}

File: modules/shared/src/main/java/io/simplesource/saga/shared/utils/StreamAppUtils.java
package io.simplesource.saga.shared.utils;

import io.simplesource.kafka.spec.TopicSpec;
import io.simplesource.saga.shared.topics.TopicCreation;
import org.apache.kafka.clients.admin.AdminClient;
import org.apache.kafka.clients.admin.CreateTopicsResult;
import org.apache.kafka.clients.admin.NewTopic;
import org.apache.kafka.streams.KafkaStreams;
import org.apache.kafka.streams.StreamsConfig;
import org.apache.kafka.streams.Topology;

import java.util.List;
import java.util.Properties;
import java.util.Set;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public final class StreamAppUtils {
    public static final CreateTopicsResult addMissingTopics(AdminClient adminClient, List<TopicCreation> topics){
        try {
            Set<String> existingTopics = adminClient.listTopics().names().get();
            Stream<NewTopic> newTopics = topics.stream()
                    .filter(t -> !existingTopics.contains(t.topicName))
                    .map(t -> {
                        TopicSpec spec = t.topicSpec;
                        return new NewTopic(t.topicName, spec.partitionCount(), spec.replicaCount()).configs(spec.config());
                    });
            return adminClient.createTopics(newTopics.collect(Collectors.toList()));
        } catch (Exception e) {
            throw new RuntimeException("Unable to add missing topics");
        }
    }

    public static void runStreamApp(Properties config, Topology topology) {
        KafkaStreams streams = new KafkaStreams(topology, new StreamsConfig(config));

        streams.cleanUp();
        streams.start();

        Runtime.getRuntime().addShutdownHook(new Thread(streams::close));
    }

}
